design/merge-patch-and-strategic-in-resource-modifier.md

```diff
@@ -0,0 +1,193 @@
+    - [User Stories](#user-stories)
+        - [Scenario 1](#scenario-1)
+        - [Scenario 2](#scenario-2)
+    - [Detailed Design](#detailed-design)
+        - [How to choose the right patch type](#how-to-choose-the-right-patch-type)
+        - [New Field MergePatches](#new-field-mergepatches)
+        - [New Field StrategicPatches](#new-field-strategicpatches)
+        - [Conditional Patches in ALL Patch Types](#conditional-patches-in-all-patch-types)
+        - [Wildcard Support for GroupResource](#wildcard-support-for-groupresource)
+        - [Helper Command to Generate Merge Patch and Strategic Merge Patch](#helper-command-to-generate-merge-patch-and-strategic-merge-patch)
+    - [Security Considerations](#security-considerations)
+    - [Compatibility](#compatibility)
+    - [Implementation](#implementation)
+    - [Future Enhancements](#future-enhancements)
+    - [Open Issues](#open-issues)
+
+## Abstract
+Velero introduced the concept of Resource Modifiers in v1.12.0. This feature allows the user to specify a configmap with a set of rules to modify the resources during restore. The user can specify the filters to select the resources and then specify the JSON Patch to apply on the resource. This feature is currently limited to the operations supported by JSON Patch RFC.
+This proposal is to add support for JSON Merge Patch and Strategic Merge Patch in the Resource Modifiers. This will allow the user to use the same configmap to apply JSON Merge Patch and Strategic Merge Patch on the resources during restore.
+
+## Goals
+- Allow the user to specify a JSON patch, JSON Merge Patch or Strategic Merge Patch for modification.
+- Allow the user to specify multiple JSON Patch, JSON Merge Patch or Strategic Merge Patch.
+- Allow the user to specify mixed JSON Patch, JSON Merge Patch and Strategic Merge Patch in the same configmap.
+
+## Non Goals
+- Deprecating the existing RestoreItemAction plugins for standard substitutions(like changing the namespace, changing the storage class, etc.)
+
+## User Stories
+
+### Scenario 1
+- Alice has some Pods and part of them have an annotation `{"for": "bar"}`.
+- Alice wishes to restore these Pods to a different cluster without this annotation.
+- Alice can use this feature to remove this annotation during restore.
+
+### Scenario 2
+- Bob has a Pod with several containers and one container with name nginx has an image `repo1/nginx`.
+- Bob wishes to restore this Pod to a different cluster, but new cluster can not access repo1, so he pushes the image to repo2.
+- Bob can use this feature to update the image of container nginx to `repo2/nginx` during restore.
+
+## Detailed Design
+- The design and approach is inspired by kubectl patch command and [this doc](https://kubernetes.io/docs/tasks/manage-kubernetes-objects/update-api-object-kubectl-patch/).
+- New fields `MergePatches` and `StrategicPatches` will be added to the `ResourceModifierRule` struct to support all three patch types.
```

Comment:
Suggest adding a new field `type` to distinguish the different patch types, `JSONPatch` is the default value if not specified. In this way, the newly added patch types share a similar format with the current solution.

```
version: v1
resourceModifierRules:
- conditions:
    groupResource: pods
    namespaces:
    - ns1
  type: mergepatch
  patches:
  - |
    {
      "metadata": {
        "annotations": {
          "foo": null
        }
      }
    }
```

Reply:
the origin design is adding a PatchType field, but @anshulahuja98 think three arrays is better, see the discussion https://kubernetes.slack.com/archives/C021GPR1L3S/p1694084640935609?thread_ts=1693808146.936179&cid=C021GPR1L3S

Reply:
I'm fine with either approach

Reply:
I am wondering if we go with this approach (using same field) - how do we ensure compatibility with current contracts. And also only permit usage of only 1 patch type potentially
Can you try to see how it can be extended @27149chen ?

Having 3 different arrays ensure that we don't  break current working.

And in future let's say we have say jq based patches, we can add another array without editing current contracts.
This is just my initial thought, I am open to discussion.

Reply:
Let's use the 3-arrays approach if we have decided.